sql/core/src/main/scala/org/apache/spark/sql/execution/stat/StatFunctions.scala

```diff
@@ -246,10 +246,15 @@ object StatFunctions extends Logging {
       if (stats.endsWith("%")) {
         val index = percentileIndex
         percentileIndex += 1
-        (child: Expression) =>
+        (child: Expression) => if (child.dataType.isInstanceOf[NumericType]) {
           GetArrayItem(
-            new ApproximatePercentile(child, Literal.create(percentiles)).toAggregateExpression(),
```

Comment:
did we fail for mismatched data type before when the `child` is not numeric type?

Reply:
With Implicitcast we can support child type =(numeric|string), shall we add string back?

Reply:
```
select approx_percentile(v, array(0.2, 0.6, 0.8)) from values ('string'), ('1.234'), ('1 day') t(v);
[1.234,1.234,1.234]
```

Reply:
also date timestamp are expected

Reply:
Now we never fail, right? If `child` is not numeric, we create null literal. What's the behavior before?

Reply:
We will fail with type check as before. now I have add StringType back, the behavior for chid type is exactly the same.

Reply:
### the result of master branch today
```sql
spark-sql> select approx_percentile(1, array(0, 1, '1a')) from values ('string'), ('1'), ('1 day') t(v);
[1,1,1]
Time taken: 0.095 seconds, Fetched 1 row(s)
spark-sql> select approx_percentile(1, array(0, 1, '1a'));
[1,1,1]
Time taken: 0.097 seconds, Fetched 1 row(s)
spark-sql> select approx_percentile('1', array(0, 1, '1a'));
[1.0,1.0,1.0]
Time taken: 0.072 seconds, Fetched 1 row(s)
spark-sql> select approx_percentile(date '1111-11-11', array(0, 1, '1a'));
[1111-11-04,1111-11-04,1111-11-04]
Time taken: 0.075 seconds, Fetched 1 row(s)
spark-sql> select approx_percentile(timestamp'1111-11-11', array(0, 1, '1a'));
[1111-11-03 23:54:17.0,1111-11-03 23:54:17.0,1111-11-03 23:54:17.0]
Time taken: 0.067 seconds, Fetched 1 row(s)
spark-sql> select approx_percentile(true, array(0, 1, '1a'));
Error in query: cannot resolve 'percentile_approx(true, array(CAST(0 AS STRING), CAST(1 AS STRING), '1a'), 10000)' due to data type mismatch: argument 1 requires (numeric or date or timestamp) type, however, 'true' is of boolean type.; line 1 pos 7;
'Aggregate [unresolvedalias(percentile_approx(true, cast(array(cast(0 as string), cast(1 as string), 1a) as array<double>), 10000, 0, 0), None)]
+- OneRowRelation

spark-sql> select approx_percentile(X'AB', array(0, 1, '1a'));
Error in query: cannot resolve 'percentile_approx(X'AB', array(CAST(0 AS STRING), CAST(1 AS STRING), '1a'), 10000)' due to data type mismatch: argument 1 requires (numeric or date or timestamp) type, however, 'X'AB'' is of binary type.; line 1 pos 7;
'Aggregate [unresolvedalias(percentile_approx(0xAB, cast(array(cast(0 as string), cast(1 as string), 1a) as array<double>), 10000, 0, 0), None)]
+- OneRowRelation
```

Reply:
Beyond this PR, the date/timestamp results are right?

Reply:
Do I need to add some script SQL test for this?

Reply:
### Here is the result of this PR, but with out one of the invaild percentage value
```sql
+-- !query 12
+select approx_percentile(1, array(0, 1))
+-- !query 12 schema
+struct<percentile_approx(1, array(0, 1), 10000):array<int>>
+-- !query 12 output
+[1,1]
+
+
+-- !query 13
+select approx_percentile('1', array(0, 1))
+-- !query 13 schema
+struct<percentile_approx(CAST(1 AS DOUBLE), array(0, 1), 10000):array<double>>
+-- !query 13 output
+[1.0,1.0]
+
+
+-- !query 14
+select approx_percentile(date '1111-11-11', array(0, 1))
+-- !query 14 schema
+struct<percentile_approx(DATE '1111-11-11', array(0, 1), 10000):array<date>>
+-- !query 14 output
+[1111-11-04,1111-11-04]
+
+
+-- !query 15
+select approx_percentile(timestamp'1111-11-11', array(0, 1))
+-- !query 15 schema
+struct<percentile_approx(TIMESTAMP '1111-11-11 00:00:00', array(0, 1), 10000):array<timestamp>>
+-- !query 15 output
+[1111-11-03 23:52:58.0,1111-11-03 23:52:58.0]
+
+
+-- !query 16
+select approx_percentile(true, array(0, 1))
+-- !query 16 schema
+struct<>
+-- !query 16 output
+org.apache.spark.sql.AnalysisException
+cannot resolve 'percentile_approx(true, array(0, 1), 10000L)' due to data type mismatch: argument 1 requires (numeric or date or timestamp) type, however, 'true' is of boolean type.; line 1 pos 7
+
+
+-- !query 17
+select approx_percentile(X'AB', array(0, 1))
+-- !query 17 schema
+struct<>
+-- !query 17 output
+org.apache.spark.sql.AnalysisException
+cannot resolve 'percentile_approx(X'AB', array(0, 1), 10000L)' due to data type mismatch: argument 1 requires (numeric or date or timestamp) type, however, 'X'AB'' is of binary type.; line 1 pos 7
```